cpg_pipes/pipeline/pipeline.py

```diff
@@ -0,0 +1,956 @@
+        wait: bool | None = False,
+    ) -> Any | None:
+        """
+        Submits Hail Batch jobs.
+        """
+        if dry_run is None:
+            dry_run = self.dry_run
+        
+        self.set_stages(self._stages_in_order)
+
+        if self.b:
+            logger.info(f'Will submit {self.b.total_job_num} jobs:')
+            for label, stat in self.b.labelled_jobs.items():
+                logger.info(
+                    f'  {label}: {stat["job_n"]} for {len(stat["samples"])} samples'
+                )
+            logger.info(f'  Other jobs: {self.b.other_job_num}')
+
+            return self.b.run(
+                dry_run=dry_run,
+                delete_scratch_on_exit=not self.keep_scratch,
+                wait=wait,
+            )
+        shutil.rmtree(self.local_tmp_dir)
+        return None
+
+    def _validate_first_last_stage(self) -> tuple[int | None, int | None]:
+        # Validating the --first-stage and --last-stage parameters
+        stage_names = list(self._stages_dict.keys())
+        lower_stage_names = [s.lower() for s in stage_names]
+        first_stage_num = None
+        if self.first_stage:
+            if self.first_stage.lower() not in lower_stage_names:
+                logger.critical(
+                    f'Value for --first-stage {self.first_stage} '
+                    f'not found in available stages: {", ".join(stage_names)}'
+                )
+            first_stage_num = lower_stage_names.index(self.first_stage.lower())
+        last_stage_num = None
+        if self.last_stage:
+            if self.last_stage.lower() not in lower_stage_names:
+                logger.critical(
+                    f'Value for --last-stage {self.last_stage} '
+                    f'not found in available stages: {", ".join(stage_names)}'
+                )
+            last_stage_num = lower_stage_names.index(self.last_stage.lower())
+        return first_stage_num, last_stage_num
+
+    def set_stages(self, stages_classes: list[StageDecorator]):
+        """
+        Iterate over stages and call add_to_the_pipeline() on each.
+        Effectively creates all Hail Batch jobs through Stage.queue_jobs().
+        """
+        # Initializing stage objects
+        stages = [cls(self) for cls in stages_classes]
+        for stage_ in stages:
+            if stage_.name in self._stages_dict:
+                raise ValueError(
+                    f'Stage {stage_.name} is already defined. Check your '
+                    f'list for duplicates: {", ".join(s.name for s in stages)}'
+                )
+            self._stages_dict[stage_.name] = stage_
+
+        first_stage_num, last_stage_num = self._validate_first_last_stage()
+
+        # First round - checking which stages we require, even if they are skipped.
+        # If there are required stages that are not defined, we defined them
+        # into `additional_stages` as `skipped`.
+        # TODO: use TopologicalSorter
```

Comment:
Is this a remnant that could now be addressed with Python 3.10 being supported?

Reply:
That's just a reminder to myself. Removed it!